### ISSUE TITLE
Problem with buffer cleanup/release (Code: MEDIA_ERR_APPEND_SOURCEBUFFER)

### ISSUE THREAD

**Issue by qchroman** [Orange-OpenSource/hasplayer.js#70]
Hello all,

I may have found an issue with buffers. It is simple to replicate. Open the page with your test player (I have used [1.2.6 player's test page](https://orange-opensource.github.io/hasplayer.js/1.2.6/dashif.html)) and put this link inside

https://djkxw9frc3oss.cloudfront.net/ss/unencrypted/ed_h264_720p_multiple-audio-tracks/ElephantsDream.ism/Manifest

Play it for 11 minutes or so (till you will see movie credits) then seek back to 5th second, for example (I tried seeking to a black area on the seek bar which is closed to the beginning of the movie). The player stops and gives this error message:

> ...
> [16:12:11.678]<video> timeupdate event: 16.910249 
> [16:12:11.680][BufferController][video] Working time: 16.910249 (paused = false) 
> [16:12:11.681][BufferController][video] Buffer level = 1.0897499999999987 (time:16.910249) 
> [16:12:11.688][BufferController][audio] Working time: 16.930249 (paused = false) 
> [16:12:11.689][BufferController][audio] Buffer level = 577.663128 (time:16.930249) 
> [16:12:11.694][BufferController][text] Working time: 16.930249 (paused = false) 
> [16:12:11.695][BufferController][text] Buffer level = 583.069751 (time:16.930249) 
> [16:12:11.707][FragmentLoader][video] Loaded: https://djkxw9frc3oss.cloudfront.net/ss/unencrypted/ed_h264_720p_multiple-a…racks/ElephantsDream.ism/QualityLevels(2962000)/Fragments(video=5960000000) (200, 223ms, 366ms) 
> [16:12:11.709][BufferController][video] Media loaded  https://djkxw9frc3oss.cloudfront.net/ss/unencrypted/ed_h264_720p_multiple-a…racks/ElephantsDream.ism/QualityLevels(2962000)/Fragments(video=5960000000) 
> [16:12:11.711][BufferController][video] Buffer segment from url  https://djkxw9frc3oss.cloudfront.net/ss/unencrypted/ed_h264_720p_multiple-a…racks/ElephantsDream.ism/QualityLevels(2962000)/Fragments(video=5960000000) 
> [16:12:11.713][BufferController][video] Buffering segment 
> ERROR: {"type":"error","data":{"code":"MEDIA_ERR_APPEND_SOURCEBUFFER","message":"Failed to append data into video source buffer","data":{"code":22,"message":"QuotaExceededError","data":"Failed to execute 'appendBuffer' on 'SourceBuffer': The SourceBuffer is full, and cannot free space to append additional buffers."}}}
> [16:12:11.716][StreamController] Reset 
> [16:12:11.716][Stream] Do pause. 
> [16:12:11.718][DRM][PM_21Jan2015] Teardown 
> [16:12:11.719][Stream] Reset 
> [16:12:11.719][Stream] Do pause. 
> [16:12:11.722][Error] Code: MEDIA_ERR_APPEND_SOURCEBUFFER, Message: Failed to append data into video source buffer, Data: {
> 	"code": 22,
> 	"message": "QuotaExceededError",
> 	"data": "Failed to execute 'appendBuffer' on 'SourceBuffer': The SourceBuffer is full, and cannot free space to append additional buffers."
> } 
> [16:12:11.722][BufferController][video] STOP 
> [16:12:11.727][BufferController][audio] STOP 
> [16:12:11.728][BufferController][text] STOP 

Same happens [here](https://orange-opensource.github.io/hasplayer.js/1.2.6/player.html?url=https://djkxw9frc3oss.cloudfront.net/ss/unencrypted/ed_h264_720p_multiple-audio-tracks/ElephantsDream.ism/Manifest). You wait for credits, don't skeep right to them, watch a full movie. Then seek to som 10th or 5th second - player shows buffering animation, buffers content as seen on Networks tab, but playback never starts.

Testing on Windows 10, Chrome (47.0.2526.106 m)

Thank you.

**Comment by qchroman** [comment]
Sorry, were you able to look into this? It seems that now the IE and Chrome are crashing if you seek back as described above.

**Comment by qchroman** [comment]
Hello all, 

anyone has any idea why this happens? I mean, I tried removing the source buffer after you jump outside buffered zone, but, for some reason it's either not removed or not removed quick enough - the playback still fails with same error message. Is it a bug in the player, browser? Kind of stuck with this nasty issue :(

Thank you.

**Comment by inbre001** [comment]
we had a simular error ... with a previous version.
But since 1.2.6 this seemed to have been fixed.  
i'll try to replicate your error tomorrow and report back then.

**Comment by qchroman** [comment]
@inbre001, thank you very much for your response. I will eagerly wait for a result. If you require anything from my side to help you debug, please, let me know.

Thank you.

**Comment by bbert** [comment]
Hi,
I confirm the problem you get, even if did not manage to get the quota exceeded error. But I notice that the browser may remove itself some buffer parts, maybe to avoid some quota exceeded error.
Then when you seek to some buffer part that the browser removed itself, the player still consider the corresponding fragments have already been loaded, and do nothing.
We have to reconsider the the FragmentModel's function isFragmentLoadedOrPending(). I will let you know.

**Comment by inbre001** [comment]
@qchroman : Under Google Chrome 49.0.2623.28 beta i  could not replicate the Quota Exceeded error anymore. I'll try to replicate it again tomorrow on a different Setup.

@bbert: Do i understand correctly that the Player might have to re-verify what is still actually  available in the Buffer, because the Browser might have removed some parts to free memory? - Re Evaluating the entire buffer is hopefully not required, maybe their is an Event/Status to listen on which can be used to  updated the Player information about removed parts?

**Comment by bbert** [comment]
Hi,
No, we better have to re-evaluating the FragmentModel that keeps in memory the already loaded fragments requests.

**Comment by qchroman** [comment]
Hello @bbert, just wanted to check whether you have a fix on your roadmap or it is something I could look at on my own.

**Comment by bbert** [comment]
Hi,
Yes, it is on our roadmap, I hope fix would be available by end of february.
But you can have a look on your side. 
It can be very simple since the idea is to not check anymore if fragment has been already loaded by FragmentModel (self.fragmentController.isFragmentLoadedOrPending() in BufferController.js).
But need to check if it still works in all cases.

**Comment by qchroman** [comment]
Thank you @bbert. By doing VERY rude thing and simply commenting out the executedRequests.push(request); inside the fragment model I was able to seek back to removed part of the buffer and playback didn't stop for me. But, the loaded part (this white line) of the seek bar becomes very small and ends right next to the seeker which makes me believe that the buffer is full and player is constantly removing old chunks or something like that. I will try to figure this out on my own, of course, but, will be eagerly waiting for the official release.

Thank you, again.

**Comment by qchroman** [comment]
One more thing I have noticed. The dash.js player cleans up after himself i.e. when the seekbar (that white bar) starts shrinking on the left side (testing in Chrome, I don't see the same behavior of seekbar in IE), the dash.js starts cleaning out the buffer. And I don't think this happens in hasplayer.js. Please, correct me if I'm mistaken, but removeBuffer or clearBuffer or remove method of the sourceBufferExt is never called during the playback. Is it suppose to be like that and dash.js simply has different buffer keeping logic?

**Comment by qchroman** [comment]
Ok, so, I figured, the problem with buffer happens only after I seek. So, as a very quick and, may be, not so elegant solution I did this in doSeek of the BufferController

  try {
    if (buffer) {
      var ranges = this.sourceBufferExt.getAllRanges(buffer);
      var rangesRemove = [];

      for(var i = 0; i < ranges.length; i++) {
        var start = ranges.start(i);
        var end = ranges.end(i);

        // Out of range
        if (time <= start || time >= end)
          removeBuffer.call(self, start, end)
      }
    }
  } catch(err) {}

I would need to add checks and, probably, first removeBuffer and _then_ call doStart. But, for now, it works for me. I seek outside the white part of the seek bar, the buffer gets removed and playback starts from a new place buffering new chunks.

**Comment by bbert** [comment]
Hi,
Regarding dash.js, we have not followed closely the developments during the last months.
I've tested last version of dash.js and yes in some cases it seems to clear the buffer when seeking.

However, for VOD streams, I would not be in favor of clearing the buffer when we do seek, thus causing re-downloading of segments previously buffered (when seeking back to an already played portion of the stream).

Maybe the best solution is to clear only the FragmentModel when seeking, i.e. to remove all the stored executed requests. Then the BufferController will not ask for downloading segments already in the buffer, but if the browser has cleared (silently) the buffer, the BufferController will ask for (re) downloading the segments, and then the FramgentModel will not state that the fragment has already been loaded.

This FragmentModel's feature was initially implemented to avoid downloading twice a segment when parallel segments downloading was used. Since in hasplayer.js we have simplified and therefore consolidated the downloading and buffering process, the <code>isFragmentLoadedOrPending()</code> function is now only used to detect we have reach the end of the available segment list (and thus end of stream for VOD streams).

I do some more tests and if it is ok with this solution I will update the source code and make a PR.
